### PR TITLE
[Plugin] Decoupled plugin register flow

### PR DESCRIPTION
**Pull request by CaptainVincent**
Fix below use case (main program static link to wasmedge.a but still need load plugins from .so)

Test the solution in
https://github.com/CaptainVincent/runwasi/actions/runs/6305204425/job/17118150750

- Class diagram

- Memory Layout

### CODE REVIEW

**Comment by juntao** [comment]
Hello, I am a [code review bot](https://github.com/flows-network/github-pr-summary/) on [flows.network](https://flows.network/).

It could take a few minutes for me to analyze this PR. Relax, grab a cup of coffee and check back later. Thanks!

File: include/loader/shared_library.h [hunk 1]
```diff
@@ -47,6 +48,7 @@ class SharedLibrary : public std::enable_shared_from_this<SharedLibrary> {
   Expect<void> load(const std::filesystem::path &Path) noexcept;
   Expect<void> load(const AST::AOTSection &AOTSec) noexcept;
   void unload() noexcept;
+  void invoke(std::function<void(void *Handle)> cb) { cb(Handle); }
```
Comment (hydai) [inline include/loader/shared_library.h:51]: We should not use `cb`. This is unclear.

**Review by hydai — CHANGES_REQUESTED** [review]
Could you test with multiple plugins loaded at the same time to make sure the counter is working as well?

File: include/loader/shared_library.h (hunk 1 above)
Comment (CaptainVincent) [reply include/loader/shared_library.h:51]: Done. Use GetDescriptor directly from Lib->get symbol.

**Comment by CaptainVincent** [comment]
> Could you test with multiple plugins loaded at the same time to make sure the counter is working as well?

Test in
https://github.com/CaptainVincent/test-autodetect-wasmedge-plugins/actions/runs/6322826230

**Comment by hydai** [comment]
Hi @CaptainVincent 
Should we ship this PR in the 0.14.0 release?
If so, @ibmibmibm and @q82419 please review this PR.

**Comment by CaptainVincent** [comment]
> Hi @CaptainVincent Should we ship this PR in the 0.14.0 release? If so, @ibmibmibm and @q82419 please review this PR.

Sure, I think the PR merged into 0.14.0 is better. However, the current rust-sdk does not support the latest WasmEdge version (test with I manually release in my forked repository). So, after I finish rebase this PR, the previous test repository cannot be verified temporarily. Nevertheless, @ibmibmibm and @q82419 still could proceed with the review. 

BTW. @apepkuss  Will the rust-sdk also catch up with the 0.14.0 version soon?

**Review by hydai — APPROVED** [review]

**Comment by hydai** [comment]
@CaptainVincent 
Please rebase again after #3072 gets merged.

**Pull request merged**